### ISSUE TITLE
Add alternative input to this sim

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/faradays-law#174]
As part of the next publication of this sim, we would like to include alternative input. This does not support screen reader description. A lot of work has gone in to alternative input and description in this sim so there is likely very little to do.  See #169 for a review of the existing work related to this.

Also assigning to @jbphet so that he is aware that this is going to be done since he is doing work in this sim for sound and will likely be seeing through publication with sound, magnification, and alternative input. Feel free to just unasign unassign or let me know if you have any questions.

**Comment by jessegreenberg** [comment]
Sim is running well with alternative input and running without error with `?fuzzBoard`. Keyboard navigation order looks reasonable. In addition to code review, the other work required here will be to
 - Replace code related to keyboard dragging with scenery-phet's GrabDragInteraction
 - Design and add a "Keyboard Shortcuts" dialog.

**Comment by jbphet** [comment]
@jessegreenberg says that the Friction sim probably has the best example from which to draw for the keyboard-based interaction with the magnet.

**Comment by jbphet** [comment]
I've added keyboard nav, including space bar pickup up, and the hints to support it.  There were custom key behaviors already implemented for the keys '1', '2', and '3'.  @emily-phet - please review the behavior in the dev version linked below and let me know if you think changes are needed.  We can also review in the sound design meeting tomorrow (9/22/2020).

dev link: https://phet-dev.colorado.edu/html/faradays-law/1.4.0-dev.62/phet/faradays-law_en_phet.html

**Comment by emily-phet** [comment]
@jbphet We reviewed in sound design meeting last week, and I believe thought all was well. @terracoda had some changes re magnet pickup arrows, and keyboard help menu content, which I believe is in another issue.

**Comment by terracoda** [comment]
Yep, I put my work in #176

**Comment by terracoda** [comment]
@jbphet, apologies for putting my screen shots in the wrong issue. Copying my comments over:

Ok, I went deep. I have notes in the design document starting here: 
https://docs.google.com/document/d/1rkjdVe7-FJh17oBIFzWv16RZG9invMrirHdh5vJWafU/edit#heading=h.i9spfv8ylvtp  

I compared the visual cues with Friction which I think is most relevant. I think for that sim (Friction) we decided to NOT visually cue the WASD keys because the Arrow keys work just fine when a screen reader is not enabled. The WASD are really only relevant to screen reader users where (with JAWS) the arrow keys perform less well. 

I suggest a grab hint on keyboard focus, more like this:
![fl-grab-release-magnet](https://user-images.githubusercontent.com/1902116/94036285-b74a9580-fd9e-11ea-93f9-20924260f6c4.png)

And then we need to discuss initial visual movement cues once the magnet is grabbed. That's a bit visually complicated for me to mock up.

We should discuss if there is a need to visually cue the sliding interaction over simple movement or if we want to combine them some how.

I mocked up initial content for Movement & Auto-Slide interactions for the keyboard shortcuts dialog, and I think there is room for improvement.

**AND Follow-up comment:**

@jbphet, I didn't mock up the after grab state as it is a bit more complicated. It would be nice to bring a visual designer in on that conversation. 

But you could try starting with something like this:
![fl-grabbed-magnet](https://user-images.githubusercontent.com/1902116/94038094-a1d66b00-fda0-11ea-92b8-47a3a075c8cd.jpg)

**Comment by terracoda** [comment]
@jbphet, I'll add, the green arrows should disappear upon keyboard focus, so the through sequence with keyboard would go like this:
**Sim Loads with visual cue for mouse/touch:**
![Screen Shot 2020-10-09 at 2 32 02 PM](https://user-images.githubusercontent.com/1902116/95611312-3ba64500-0a3c-11eb-8c86-3d7a6c971868.png)

**Upon focus, green arrows disappear, and hint to grab and release appears:**
![fl-grab-release-magnet](https://user-images.githubusercontent.com/1902116/95611525-8627c180-0a3c-11eb-9307-7521c0293976.png)

**Upon successful grab, the green arrows come back and teh dotted grabbed/draggable state is indicated by the dotted pink highlight**
![fl-grabbed-magnet](https://user-images.githubusercontent.com/1902116/95611633-a788ad80-0a3c-11eb-8c2a-34a5f9ce3f49.jpg)

**Green arrows disappear upon successful drag/move**
![Screen Shot 2020-10-09 at 2 38 14 PM](https://user-images.githubusercontent.com/1902116/95612028-3c8ba680-0a3d-11eb-89e3-8801775248cc.png)

The visual hints would restart after a general Reset or page reload.

**Comment by emily-phet** [comment]
Hey @terracoda can you be more specific about what is more complicated re the "after grab state"? Did you mean it's more complicated to mockup, or that there is something complicating about the design you would like?

If it needs to look exactly like Friction, perhaps @jbphet does not need a mockup. If it needs to look different, please ask @arouinfar if she can mock something up, with some input from you so she knows what to make. 

Ideally, this is wrapped up very quickly, and this sim moves into the code refining, QA testing stage. The idea is to get it published by the end of the year.

**Comment by terracoda** [comment]
Great question, when I first looked at this, I was thinking, once grabbed we would need new arrows and possibly 123,  but the I just tried it with the green arrows, and it looks ok to me.

We could change them to pink arrows once grabbed. We did that in Friction.

And I don't know if you want to cue the 123 at all on initial grab, or just leave that to the Keyboard Shortcuts dialog.

**Comment by arouinfar** [comment]
It sounds like you might not need additional mockups, but if that changes, I'd be happy to mock something up @terracoda.

**Comment by emily-phet** [comment]
@terracoda 
Let's just indicate the 123 in the shortcuts dialog, not in the pickup/drag cue.

**Comment by jbphet** [comment]
@terracoda and @emily-phet - I've implemented the interaction between keyboard nav and visual hints as described above and folded in what we've discussed in meetings.  This is published it in a dev version (see link below).  Please review and let me know if you think further revisions are needed.

One specific question: In the version linked below, if the user first moves the magnet with the mouse, then navigates to it using the keyboard, the grab hint isn't shown.  Basically, any successful grab of the magnet - whether via mouse, touch, or keyboard - clears the flag that says that the hint should be shown.  Is this how we'd like it to behave, or should we _always_ show the grab hint the first time the magnet receives focus via keyboard nav?

Dev link: https://phet-dev.colorado.edu/html/faradays-law/1.4.0-dev.63/phet/faradays-law_en_phet.html

**Comment by emily-phet** [comment]
@terracoda and/or @jessegreenberg should confirm, but I believe every time the user first moves the magnet (between reset all presses) with the the keyboard, they should get the grab hint.

**Comment by terracoda** [comment]
@jbphet that's a great question. I'll check other sims. I think (but need to confirm) that the we need to use successful  keyboard interaction as the flag clearer because keyboard use has special instructions.

I'll take for a run and get back to you in the morning.

**Comment by jbphet** [comment]
The sound design team just looked at Friction, and in that case the hint is shown the first time the book receives focus via keyboard nav regardless of whether it had been moved via mouse or touch, so I'll change the magnet to match that behavior.

**Comment by terracoda** [comment]
@jbphet, we have some discrepancy between the published versions and non-published version of Friction and Balloons and Static Electricity where the interaction is essential to the sim.

In the published versions of F and BASE, one has to make a successful movement with the the keyboard before the visual grab hint goes away.

However, on master for F and BASE any successful movement will make the hint go away.

In GFL, the ruler is not an essential object. In the published version of that sim, any successful movement makes the visual grab hint go away.

@jbphet, is there a built-in option for these 2 scenarios, i.e.: 
- successful keyboard interaction - hide visual grab hint
- any successful movement - hide visual grab hint?

The visual grab hint was designed specifically alternative input where we are assuming the user is using a keyboard (or some other form of alternative input, e.g. a switch) and they have vision, so not using a screen reader. Since these objects are grabbed with a button interaction, bu they do not look like buttons, the visual user needs to know that they have to activate the grab button before they can actually drag or move the object.

Therefore, I think we should only hide the visual grab hint after there is successful interaction with the keyboard. I define successful interaction as a successful move/drag.

Imagine a scenario where 2 students are collaborating. The mouse user tries things out first, then their partner tries with the keyboard but gets no hint, so doesn't know how to grab the object. They would with have to ask the teacher who may or may not know, or go to the keyboard shortcuts dialog to get instructions.

 I think hiding the visual grab hint after successful interaction with the keyboard provides the most equitable experience.

**Comment by emily-phet** [comment]
@terracoda Agreed - I think the grab hint should only be hid after successful interaction with keyboard appears. @jbphet had planned to do this based on our meeting yesterday, so no change of plans there. 

Regarding how to make this consistent across sims, I think that's a question for @jessegreenberg.

**Comment by terracoda** [comment]
I think it is ok to have a built-in option, but I can't think of a reason right now why we would want to hide the hint prematurely before successful keyboard use.

Maybe in GFL, since moving the ruler is completely optional, it's okay, but still there is no harm in being consistent there as well.

**Comment by terracoda** [comment]
It's possible, when working on the accessible mobile grab and drag we adjusted the situation unintentionally?

**Comment by terracoda** [comment]
@jbphet, I missed your https://github.com/phetsims/faradays-law/issues/174#issuecomment-713149602
Yes, follow Friction's published pattern.

**Comment by jessegreenberg** [comment]
https://github.com/phetsims/scenery-phet/issues/636 is ready for review and I believe addresses this. I tested it in faradays-law and believe that it is working as requested now. @jbphet can you please confirm?

**Comment by jbphet** [comment]
> I tested it in faradays-law and believe that it is working as requested now. @jbphet can you please confirm?

Yep, it's working great.  Thanks @jessegreenberg!

**Comment by jbphet** [comment]
@terracoda and @emily-phet - I think the behavior now matches what was described above.  Can you please review and then assign this back to me?

**Comment by emily-phet** [comment]
This looks good to me!

**Comment by terracoda** [comment]
The alternative input cues look good to me as well.

**Comment by terracoda** [comment]
I'll go ahead and close. @jbphet, please re-open if I have closed this issue prematurely.

**Comment by terracoda** [comment]
After closing, I noticed this comment https://github.com/phetsims/faradays-law/issues/109#issuecomment-413231548 which included this image from Jon Hung's documentation.

Just adding it here for reference in case we choose to make future updates.

![Screen Shot 2020-10-28 at 11 57 46 AM](https://user-images.githubusercontent.com/1902116/97449774-fbd4ce00-1914-11eb-83d4-4020e18aa321.png)

**Comment by jbphet** [comment]
Re-opening, since I want it around for QA to make updates if they need to.  We will close it when the sim has been deployed with this feature.

**Comment by muedli** [comment]
Keyboard nav is working well in 1.4.0-rc.1.

**Comment by KatieWoe** [comment]
Looks good. I did notice that the arrows don't come back when you use keyboard nav if you have already drug the magnet with mouse. This is probably intended but I wanted to check. @jbphet you can close if this is the case.

**Comment by jbphet** [comment]
@KatieWoe - Yes, the behavior that you describe in the comment immediately above is intentional.  Closing.
